Add 4.6-dev1 blog post (#1180)
godotengine/godot-website: collections/_article/dev-snapshot-godot-4-6-dev-1.md

---
title: "Dev snapshot: Godot 4.6 dev 1"
excerpt: The calm before the storm…
categories: [pre-release]
author: Thaddeus Crews
image: /storage/blog/covers/dev-snapshot-godot-4-6-dev-1.webp
image_caption_title: "RAM: Random Access Mayhem"
image_caption_description: A game by Xylem Studios
date: 2025-09-29 12:00:00
---

> The first development snapshot for 4.6 arrives at last! As is often the case for our first development snapshot, a significant portion of quality PRs from our backlog are finally able to see the light of day, as they were either locked out from the feature freeze or deemed too risky to merge for the stable release. In saying that though, this is *by far* the biggest our backlog has ever been, so getting it all in for an initial snapshot was unrealistic. As such, while this may mean a slower trickle of PRs initially, you can expect future development snapshots to further expand on the foundation that this release sets.

Reviewer: ```suggestion
The first development snapshot for 4.6 has arrived. As is often the case for our first development snapshot, a significant portion of quality PRs from our backlog are finally able to see the light of day, as they were either locked out from the 4.5 feature freeze or deemed too risky to merge for the stable release. In saying that though, this is *by far* the biggest our backlog has ever been, so getting it all in for an initial snapshot was unrealistic. As such, while this may mean a slower trickle of PRs initially, you can expect future development snapshots to further expand on the foundation that this release sets.
```

"at last" seemed odd when 4.5 has only just been released! I realise as a dev it may have been a long wait because of the feature freeze, but to an outsider it seems an odd phrase.

> The overwhelming majority of changes this time around are bugfixes, with most already slated to be backported to 4.5.1-stable in the very near future! In a sense, this release is serving not only as a foundation for 4.6, but for 4.5.1 as well; meaning, testing this build is crucial to ensure a smooth release for both. However, we've already got quite a few enhancements and features integrated that'll remain 4.6 exclusive, so those wishing to stick with 4.5 may want to hold out for 4.5.1-rc1 coming later this week.

Reviewer: ```suggestion
The overwhelming majority of changes this time around are bugfixes, with most already slated to be backported to 4.5.1-stable in the very near future! In a sense, this release is serving not only as a foundation for 4.6, but for 4.5.1 as well; meaning, testing this build is crucial to ensure a smooth release for both. However, we've already got quite a few enhancements and features integrated that will remain exclusive to 4.6, so those wishing to stick with 4.5 may want to hold out for 4.5.1-rc1 coming later this week.
```

> As usual, safety precautions should be taken with any pre-release environment. While we prepare these snapshots with the intent to b e suitable for general testing, there will always be a non-zero risk of data loss/corruption. Creating backups before hand and/or utilizing version control are strongly recommended!

Reviewer 2: ```suggestion
As usual, safety precautions should be taken with any pre-release environment. While we prepare these snapshots with the intent to be suitable for general testing, there will always be a non-zero risk of data loss/corruption. Creating backups before hand and/or utilizing version control are strongly recommended!
```

Please consider [supporting the project financially](#support), if you are able. Godot is maintained by the efforts of volunteers and a small team of paid contributors. Your donations go towards sponsoring their work and ensuring they can dedicate their undivided attention to the needs of the project.

> [Jump to the **Downloads** section](#downloads), and give it a spin right now, or continue reading to learn more about improvements in this release. You can also try the [**Web editor**](https://editor.godotengine.org/releases/4.6.dev1/), the [**XR editor**](https://www.meta.com/s/h9JcJGHfg), or the [**Android editor**](https://play.google.com/store/apps/details?id=org.godotengine.editor.v4) for this release. If you are interested in the latter, please request to join [our testing group](https://groups.google.com/g/godot-testers) to get access to pre-release builds.

Reviewer 2: ```suggestion
[Jump to the **Downloads** section](#downloads), and give it a spin right now, or continue reading to learn more about improvements in this release. You can also try the [**Web editor**](https://editor.godotengine.org/releases/4.6.dev1/), the [**XR editor**](https://www.meta.com/s/3yJ7i8kop), or the [**Android editor**](https://play.google.com/store/apps/details?id=org.godotengine.editor.v4) for this release. If you are interested in the latter, please request to join [our testing group](https://groups.google.com/g/godot-testers) to get access to pre-release builds.
```

---

*The cover illustration is from* [**RAM: Random Access Mayhem**](https://store.steampowered.com/app/2256450/RAM_Random_Access_Mayhem/?curator_clanid=41324400), *a rougelike where you play as the enemies. You can buy the game or try out the demo on [Steam](https://store.steampowered.com/app/2256450/RAM_Random_Access_Mayhem/?curator_clanid=41324400), and follow the developers on [Twitter](https://twitter.com/Xylem_Studios) or [YouTube](https://www.youtube.com/@xylemstudios).*

## Highlights

### Drag-and-drop `@export` variables

Part of starting out slow with enhancements in the development cycle means that you can expect quite a lot of QOL additions in the near future. One such addition that we're excited to showcase comes from [fkeyz](https://github.com/fkeyzuwu): the ability to drag-and-drop objects to the script editor to automatically create an exported variable! ([GH-106341](https://github.com/godotengine/godot/pull/106341))

<video autoplay loop muted playsinline title="Drag-and-drop demonstration">
  <source src="/storage/blog/dev-snapshot-godot-4-6-dev-1/drag-and-drop-export.mp4?1" type="video/mp4">
</video>

### OpenXR: Add support for spatial entities extension

> As [Bastiaan Olij](https://github.com/BastiaanOlij) notes in his PR [GH-107391](https://github.com/godotengine/godot/pull/107391), [The OpenXR Spatial entities extension](https://www.khronos.org/blog/openxr-spatial-entities-extensions-released-for-developer-feedback) was introduced to standardize obtaining and interacting with information about the user's real world environment. This is an absolute goliath of a specification, and is reflected in the implementation seeing over **75 hundred lines of code** changed. If you're interested in seeing the changes firsthand (and have the necessary equipment for it), be sure to check out Bastiaan's accompanying [demo project](https://github.com/BastiaanOlij/spatial-entities-demo).

Reviewer: ```suggestion
As [Bastiaan Olij](https://github.com/BastiaanOlij) notes in his PR [GH-107391](https://github.com/godotengine/godot/pull/107391), the [OpenXR Spatial entities extension](https://www.khronos.org/blog/openxr-spatial-entities-extensions-released-for-developer-feedback) was introduced to standardize obtaining and interacting with information about the user's real world environment. This is an absolute goliath of a specification, and is reflected in the implementation seeing over **75 hundred lines of code** changed. If you're interested in seeing the changes firsthand (and have the necessary equipment for it), be sure to check out Bastiaan's accompanying [demo project](https://github.com/BastiaanOlij/spatial-entities-demo).
```

> As [Bastiaan Olij](https://github.com/BastiaanOlij) notes in his PR [GH-107391](https://github.com/godotengine/godot/pull/107391), the [OpenXR Spatial entities extension](https://www.khronos.org/blog/openxr-spatial-entities-extensions-released-for-developer-feedback) was introduced to standardize obtaining and interacting with information about the user's real world environment. This is an absolute goliath of a specification, and is reflected in the implementation seeing over **75 hundred lines of code** changed. If you're interested in seeing the changes firsthand (and have the necessary equipment for it), be sure to check out Bastiaan's accompanying [demo project](https://github.com/BastiaanOlij/spatial-entities-demo).

Reviewer: ```suggestion
### OpenXR: Add support for Spatial Entities Extensions
As [Bastiaan Olij](https://github.com/BastiaanOlij) notes in his PR [GH-107391](https://github.com/godotengine/godot/pull/107391), the [OpenXR Spatial Entities Extensions](https://www.khronos.org/blog/openxr-spatial-entities-extensions-released-for-developer-feedback) were introduced to standardize obtaining and interacting with information about the user's real world environment. This is an absolute goliath of a specification, and is reflected in the implementation seeing over **75 hundred lines of code** changed. If you're interested in seeing the changes firsthand (and have the necessary equipment for it), be sure to check out Bastiaan's accompanying [demo project](https://github.com/BastiaanOlij/spatial-entities-demo).
```
I believe that's the proper name, according to the Khronos link.

### Hide `Control` focus when given via mouse input

> Courtesy of [Michael Alexsander](https://github.com/YeldhamDev), the focus state logic for mouse/touch is now decoupled from keyboard/joypad [GH-110250](https://github.com/godotengine/godot/pull/110250). While it's common for programs to have significant overlap between registering inputs of these types, it's not uncommon for systems to deliberately stylize the two styles separately, often handling their inputs in entirely separate ways; this change enables that granular control for toolmakers and UI designers. Included in the PR is a comprehensive ruleset for when focus is shown, which we've included below:

Reviewer: ```suggestion
Courtesy of [Michael Alexsander](https://github.com/YeldhamDev), the focus state logic for mouse and touch is now decoupled from keyboard and joypad ([GH-110250](https://github.com/godotengine/godot/pull/110250)). While it's common for programs to have significant overlap between registering inputs of these types, it's not uncommon for systems to deliberately stylize the two types separately, often handling their inputs in entirely separate ways. This change enables that granular control for toolmakers and UI designers. Included in the PR is a comprehensive ruleset for when focus is shown, which we've included below:
```

| Situation                                                                                            |     |
| ---------------------------------------------------------------------------------------------------- | --- |
| Clicking a `Control` with the mouse, giving it focus.                                                |     |
| Successfully switching focus via keyboard/joypad actions.                                            | ✅   |

> | Attempting to switching focus via keyboard/joypad actions but still remaining on the same `Control`. | ✅   |

Reviewer: ```suggestion
| Attempting to switch focus via keyboard/joypad actions but still remaining on the same `Control`. | ✅   |
```

| Clicking somewhere with the mouse while having a `Control` with visible focus.                       |     |
| Clicking with the mouse a visibly focused `Control` (deviates from how it works in browsers).        |     |
| Using `Control.grab_focus(true)`.                                                                    |     |

> 

Reviewer: Personally, I don't understand this table. What do the ticks mean?

Author: It's a bit obtuse, but the previous sentence outlines it a bit. I'm unsure of how to expand upon it further

Reviewer: Thanks. It doesn't matter for this blog, but maybe a video would help for the big 4.6 release notes article.

<div markdown=1 class="card card-info" style="margin-top: 1em;">

> Previous behavior can be achieved by enabling `gui/common/always_show_focus_state`.

Reviewer: ```suggestion
The previous behavior can be achieved by enabling `gui/common/always_show_focus_state`.
```

Anyway, [Robert Yevdokimov](https://github.com/ryevdokimov) flipped a switch and everything "just worked". ([GH-100876](https://github.com/godotengine/godot/pull/100876))

<video autoplay loop muted playsinline title="Swapping themes without a reset">
  <source src="/storage/blog/dev-snapshot-godot-4-6-dev-1/theme-swap-no-reset.mp4?1" type="video/mp4">
</video>

Oops.

To clarify: once-upon-a-time this reset was almost certainly necessary. Our internal logic for theming and customization is unrecognizable compared to their humble beginnings, and this is far from the first setting that was elevated to real-time support. It's common for settings that become real-time to have a PR that explicitly targets the feature in question, but it's rare for such features to have their changes exist in a vacuum. Perhaps there are other features out there, just like this, that have pre-conceived restrictions long-addressed without anyone realizing it. Until someone shows up to challenge those restrictions, we may never know.

<div markdown=1 class="card card-info" style="margin-top: 1em;">

> Theme showcased for comparison is the [Godot Minimal Theme](https://github.com/passivestar/godot-minimal-theme) by [passivestar](https://github.com/passivestar).

Reviewer: ```suggestion
The theme showcased for comparison is the [Godot Minimal Theme](https://github.com/passivestar/godot-minimal-theme) by [passivestar](https://github.com/passivestar).
```
